File: exhibits/pano_check.py
"""Asset availability + content fingerprint for the studio list API (mirrors pano-check.mjs)."""
from __future__ import annotations

import hashlib
import json
import re
from pathlib import Path

_REMOTE = re.compile(r'^(https?:|data:|blob:|//)')

# 与 pano-check.mjs 保持一致；改算法请同步改版本号（见该文件里的说明）
FINGERPRINT_VERSION = 'v1'
FINGERPRINT_CHUNK = 65536
FINGERPRINT_LENGTH = 16


def is_remote_panorama_url(path: str | None) -> bool:
    return bool(_REMOTE.match(str(path or '').strip()))


def resolve_asset_local_path(
    exhibit_dir: Path,
    asset_path: str | None,
    exhibits_root: Path | None = None,
) -> Path | None:
    p = str(asset_path or '').strip()
    if not p or is_remote_panorama_url(p):
        return None
    if p.startswith('/'):
        if not exhibits_root:
            return None
        return exhibits_root / p.lstrip('/')
    local = Path(p)
    return local if local.is_absolute() else exhibit_dir / p


def has_asset_file(
    exhibit_dir: Path,
    asset_path: str | None,
    exhibits_root: Path | None = None,
) -> bool:
    """通用资源存在性判断（模型 / 全景共用同一套路径解析规则）。"""
    p = str(asset_path or '').strip()
    if not p:
        return False
    if is_remote_panorama_url(p):
        return True
    local = resolve_asset_local_path(exhibit_dir, p, exhibits_root)
    if local is None:
        return False
    return local.is_file()


def is_site_root_pano_path(p: str | None) -> bool:
    return str(p or '').strip().startswith('/')


def check_panorama_path_availability(panorama_path: str | None, exhibits_root: Path) -> bool | str:
    """批量面板校验：True / False / 'unknown'"""
    p = str(panorama_path or '').strip()
    if not p:
        return False
    if is_remote_panorama_url(p):
        return 'unknown'
    if is_site_root_pano_path(p):
        return 'unknown'
    if p.startswith('../'):
        ref = exhibits_root / '_template'
        return has_asset_file(ref, p, exhibits_root)
    return 'unknown'


def asset_fingerprint(
    exhibit_dir: Path,
    asset_path: str | None,
    exhibits_root: Path | None = None,
) -> str:
    """本地资源内容指纹（长度 + 头尾各 64KiB）；远程 / 缺失 / 读不出一律返回 ''。"""
    local = resolve_asset_local_path(exhibit_dir, asset_path, exhibits_root)
    if local is None:
        return ''
    try:
        if not local.is_file():
            return ''
        size = local.stat().st_size
        h = hashlib.sha1()
        h.update(f'{FINGERPRINT_VERSION}|{size}|'.encode('utf-8'))
        with open(local, 'rb') as f:
            head_len = min(size, FINGERPRINT_CHUNK)
            if head_len > 0:
                h.update(f.read(head_len))
            tail_start = max(FINGERPRINT_CHUNK, size - FINGERPRINT_CHUNK)
            if size > tail_start:
                f.seek(tail_start)
                h.update(f.read(size - tail_start))
        return h.hexdigest()[:FINGERPRINT_LENGTH]
    except OSError:
        return ''


# ---------- 全景图候选清单（批量编辑的下拉选择用；与 pano-check.mjs 保持一致） ----------
PANO_EXT = {'.jpg', '.jpeg', '.png', '.webp'}
PANO_SKIP_DIRS = {'node_modules', 'vendor', '_runtime', '_dev', 'e2e', 'test-results', 'playwright-report'}
PANO_PUBLIC_DIRS = ['共享背景', '_panoramas', 'shared/panoramas']
# 等距柱状投影固有 2:1；放宽到 1.7 会把 16:9（1.778）的截图也收进来
PANO_RATIO_MIN = 1.9
PANO_RATIO_MAX = 2.1
PANO_MAX_SCAN = 400


def image_size(buf: bytes):
    """从图头读宽高（JPEG / PNG / WebP）；读不出返回 None。不解码像素。"""
    if not buf or len(buf) < 16:
        return None
    if buf[:8] == b'\x89PNG\r\n\x1a\n':
        return (int.from_bytes(buf[16:20], 'big'), int.from_bytes(buf[20:24], 'big'))
    if len(buf) >= 30 and buf[:4] == b'RIFF' and buf[8:12] == b'WEBP':
        fmt = buf[12:16]
        if fmt == b'VP8X':
            return ((int.from_bytes(buf[24:27], 'little') & 0xFFFFFF) + 1,
                    (int.from_bytes(buf[27:30], 'little') & 0xFFFFFF) + 1)
        if fmt == b'VP8 ':
            return (int.from_bytes(buf[26:28], 'little') & 0x3FFF,
                    int.from_bytes(buf[28:30], 'little') & 0x3FFF)
        if fmt == b'VP8L' and buf[20] == 0x2F:
            b = int.from_bytes(buf[21:25], 'little')
            return ((b & 0x3FFF) + 1, ((b >> 14) & 0x3FFF) + 1)
        return None
    if buf[0] == 0xFF and buf[1] == 0xD8:
        i = 2
        while i + 9 < len(buf):
            if buf[i] != 0xFF:
                i += 1
                continue
            marker = buf[i + 1]
            if marker == 0xD8 or marker == 0x01 or 0xD0 <= marker <= 0xD7:
                i += 2
                continue
            ln = int.from_bytes(buf[i + 2:i + 4], 'big')
            if 0xC0 <= marker <= 0xCF and marker not in (0xC4, 0xC8, 0xCC):
                return (int.from_bytes(buf[i + 7:i + 9], 'big'), int.from_bytes(buf[i + 5:i + 7], 'big'))
            if ln < 2:
                return None
            i += 2 + ln
    return None


def is_panorama_ratio(size) -> bool:
    if not size or not size[0] or not size[1]:
        return False
    r = size[0] / size[1]
    return PANO_RATIO_MIN <= r <= PANO_RATIO_MAX


def panorama_to_root_rel(exhibit_dir: Path, panorama_path: str | None, exhibits_root: Path | None) -> str | None:
    p = str(panorama_path or '').strip()
    if not p or is_remote_panorama_url(p):
        return None
    if p.startswith('/'):
        if not exhibits_root:
            return None
        local = exhibits_root / p.lstrip('/')
    elif Path(p).is_absolute():
        local = Path(p)
    else:
        local = exhibit_dir / p
    if not exhibits_root:
        return None
    try:
        rel = local.resolve().relative_to(exhibits_root.resolve())
    except ValueError:
        return None
    return rel.as_posix()


def _read_local_image_meta(exhibits_root: Path, root_rel: str) -> dict | None:
    full = exhibits_root.joinpath(*root_rel.split('/'))
    if full.suffix.lower() not in PANO_EXT:
        return None
    try:
        if not full.is_file():
            return None
        with open(full, 'rb') as f:
            size = image_size(f.read(65536))
    except OSError:
        return None
    if not is_panorama_ratio(size):
        return None
    return {'width': size[0], 'height': size[1]}


def _list_exhibit_dirs(exhibits_root: Path) -> list[str]:
    out: list[str] = []
    try:
        items = sorted(exhibits_root.iterdir(), key=lambda x: x.name)
    except OSError:
        return out
    for it in items:
        if not it.is_dir() or it.name.startswith('_') or it.name.startswith('.'):
            continue
        if (it / 'config.json').is_file():
            out.append(it.name)
    return out


def list_panorama_candidates(exhibits_root: Path) -> list[dict]:
    """可信来源的全景候选：config 引用 + 公共目录；path 相对 exhibits/、正斜杠。"""
    root = exhibits_root.resolve()
    merged: dict[str, dict] = {}

    def add(entry: dict) -> None:
        if len(merged) >= PANO_MAX_SCAN:
            return
        prev = merged.get(entry['path'])
        if not prev or (prev.get('source') == 'shared' and entry.get('source') == 'configured'):
            merged[entry['path']] = entry

    for ex in _list_exhibit_dirs(root):
        if len(merged) >= PANO_MAX_SCAN:
            break
        exhibit_dir = root / ex
        try:
            cfg = json.loads((exhibit_dir / 'config.json').read_text(encoding='utf-8'))
        except (OSError, json.JSONDecodeError):
            continue
        pano = (cfg.get('assets') or {}).get('panorama')
        if not pano or is_remote_panorama_url(pano):
            continue
        root_rel = panorama_to_root_rel(exhibit_dir, pano, root)
        if not root_rel:
            continue
        meta = _read_local_image_meta(root, root_rel)
        if meta:
            add({'path': root_rel, **meta, 'source': 'configured'})

    def walk_shared(d: Path, rel: str) -> None:
        if len(merged) >= PANO_MAX_SCAN:
            return
        try:
            items = sorted(d.iterdir(), key=lambda x: x.name)
        except OSError:
            return
        for it in items:
            if len(merged) >= PANO_MAX_SCAN:
                return
            if it.name.startswith('.'):
                continue
            r = f'{rel}/{it.name}' if rel else it.name
            if it.is_dir():
                if it.name not in PANO_SKIP_DIRS:
                    walk_shared(it, r)
                continue
            if not it.is_file():
                continue
            meta = _read_local_image_meta(root, r)
            if meta:
                add({'path': r, **meta, 'source': 'shared'})

    for pub in PANO_PUBLIC_DIRS:
        if len(merged) >= PANO_MAX_SCAN:
            break
        pub_dir = root.joinpath(*pub.split('/'))
        if pub_dir.is_dir():
            walk_shared(pub_dir, pub.replace('\\', '/'))

    order = {'configured': 0, 'shared': 1}
    return sorted(merged.values(), key=lambda x: (order.get(x.get('source'), 9), x['path']))

File: exhibits/pano-check.mjs
/** Whether a configured asset path (panorama, model, …) resolves to an existing local file (or remote URL). */
import crypto from 'node:crypto'
import fs from 'node:fs'
import path from 'node:path'

const REMOTE = /^(https?:|data:|blob:|\/\/)/

export function isRemotePanoramaUrl(p) {
  return REMOTE.test(String(p ?? '').trim())
}

export function resolvePanoramaLocalPath(exhibitDir, panoramaPath, exhibitsRoot = null) {
  const p = String(panoramaPath ?? '').trim()
  if (!p || isRemotePanoramaUrl(p)) return null
  if (p.startsWith('/')) {
    if (!exhibitsRoot) return null
    return path.join(exhibitsRoot, p.replace(/^\/+/, ''))
  }
  if (path.isAbsolute(p)) return p
  return path.join(exhibitDir, p)
}

/** 通用资源存在性判断（模型 / 全景 / 封面共用同一套路径解析规则） */
export function hasAssetFile(exhibitDir, assetPath, exhibitsRoot = null) {
  const p = String(assetPath ?? '').trim()
  if (!p) return false
  if (isRemotePanoramaUrl(p)) return true
  const local = resolvePanoramaLocalPath(exhibitDir, p, exhibitsRoot)
  if (!local) return false
  try {
    return fs.statSync(local).isFile()
  } catch {
    return false
  }
}

/** 批量面板校验全景路径：true / false / 'unknown'（远程或展品内相对路径） */
export function checkPanoramaPathAvailability(panoramaPath, exhibitsRoot) {
  const p = String(panoramaPath ?? '').trim()
  if (!p) return false
  if (isRemotePanoramaUrl(p)) return 'unknown'
  if (p.startsWith('../') || p.startsWith('/')) {
    const refDir = path.join(exhibitsRoot, '_template')
    return hasAssetFile(refDir, p, exhibitsRoot) ? true : false
  }
  return 'unknown'
}

/*
 * 资源内容指纹
 * ------------
 * 工作台按「哪些展品用了同一张背景」分组时，路径是靠不住的判据：
 *   · 只比文件名 → 各展品自带的 assets/panorama.jpg 会被误并成一组（其实是不同的图）；
 *   · 只比全路径 → 同一张图复制进 20 个展品目录，会被拆成 20 组（其实是同一张图）。
 * 唯一正确的判据是内容，所以由服务端算一个指纹给前端。
 *
 * 为什么不整文件哈希：全景图常有 5–8MB，100+ 件就是 1GB，每次刷新列表都读一遍太慢。
 * 改为「文件长度 + 头 64KiB + 尾 64KiB」，无论文件多大都是常数开销（≤128KiB）。
 * 代价：两张图若长度相同、首尾 64KiB 也完全相同、只在中段有差异，会被判为同一张。
 * 对 JPEG/PNG（头部含尺寸与量化表、尾部含结束标记）而言这需要刻意构造，实拍图不会发生。
 *
 * 三份服务端实现（Node / Python / PHP）必须给出同一个值，否则换个服务端分组就会重排；
 * pano-check.test.mjs 里有跨实现一致性测试。改算法请同步改 FINGERPRINT_VERSION。
 */
export const FINGERPRINT_VERSION = 'v1'
export const FINGERPRINT_CHUNK = 65536
export const FINGERPRINT_LENGTH = 16

function readChunk(fd, position, length) {
  const buf = Buffer.alloc(length)
  let got = 0
  while (got < length) {
    const n = fs.readSync(fd, buf, got, length - got, position + got)
    if (n <= 0) break            // 文件被截短：按实际读到的算
    got += n
  }
  return got === length ? buf : buf.subarray(0, got)
}

/** 本地资源的内容指纹；远程 URL / 缺失文件 / 读不出来一律返回 ''（前端按「无指纹」处理） */
export function assetFingerprint(exhibitDir, assetPath, exhibitsRoot = null) {
  const p = String(assetPath ?? '').trim()
  if (!p || isRemotePanoramaUrl(p)) return ''
  const local = resolvePanoramaLocalPath(exhibitDir, p, exhibitsRoot)
  if (!local) return ''
  let fd = null
  try {
    const st = fs.statSync(local)
    if (!st.isFile()) return ''
    const size = st.size
    const h = crypto.createHash('sha1')
    h.update(`${FINGERPRINT_VERSION}|${size}|`, 'utf8')
    fd = fs.openSync(local, 'r')
    const headLen = Math.min(size, FINGERPRINT_CHUNK)
    if (headLen > 0) h.update(readChunk(fd, 0, headLen))
    const tailStart = Math.max(FINGERPRINT_CHUNK, size - FINGERPRINT_CHUNK)
    if (size > tailStart) h.update(readChunk(fd, tailStart, size - tailStart))
    return h.digest('hex').slice(0, FINGERPRINT_LENGTH)
  } catch {
    return ''
  } finally {
    if (fd !== null) {
      try { fs.closeSync(fd) } catch { /* 已关闭或句柄失效，忽略 */ }
    }
  }
}

/*
 * 全景图候选清单
 * --------------
 * 批量编辑「全景贴图」下拉：只列可信来源，不全树扫描。
 *   1. 各展品 config 里实际声明的 assets.panorama（本地文件）；
 *   2. 约定公共目录（共享背景/、_panoramas/、shared/panoramas/）下的 2:1 图片。
 * 宽高比 1.9～2.1 仅作格式校验，不作「发现全景」的依据——2:1 贴图/横幅不能进候选。
 */
const PANO_EXT = new Set(['.jpg', '.jpeg', '.png', '.webp'])
const PANO_SKIP_DIRS = new Set(['node_modules', 'vendor', '_runtime', '_dev', 'e2e', 'test-results', 'playwright-report'])
/** 公共全景目录（相对 exhibits/ 根）；仅在这些目录内扫描未配置的 2:1 图 */
export const PANO_PUBLIC_DIRS = ['共享背景', '_panoramas', 'shared/panoramas']
// 等距柱状投影固有 2:1。放到 1.7 会把 16:9 的截图（1.778）也当成全景图 —— 宁可漏，不可错。
export const PANO_RATIO_MIN = 1.9
export const PANO_RATIO_MAX = 2.1
const PANO_MAX_SCAN = 400

/** 从图头读宽高；只支持 JPEG / PNG / WebP(VP8X|VP8|VP8L)，读不出返回 null */
export function imageSize(buf) {
  if (!buf || buf.length < 16) return null
  // PNG: 8 字节签名 + IHDR
  if (buf.readUInt32BE(0) === 0x89504e47 && buf.readUInt32BE(4) === 0x0d0a1a0a) {
    return { width: buf.readUInt32BE(16), height: buf.readUInt32BE(20) }
  }
  // WebP: RIFF....WEBP
  if (buf.length >= 30 && buf.toString('ascii', 0, 4) === 'RIFF' && buf.toString('ascii', 8, 12) === 'WEBP') {
    const fmt = buf.toString('ascii', 12, 16)
    if (fmt === 'VP8X') return { width: (buf.readUIntLE(24, 3) & 0xffffff) + 1, height: (buf.readUIntLE(27, 3) & 0xffffff) + 1 }
    if (fmt === 'VP8 ') return { width: buf.readUInt16LE(26) & 0x3fff, height: buf.readUInt16LE(28) & 0x3fff }
    if (fmt === 'VP8L' && buf[20] === 0x2f) {
      const b = buf.readUInt32LE(21)
      return { width: (b & 0x3fff) + 1, height: ((b >> 14) & 0x3fff) + 1 }
    }
    return null
  }
  // JPEG: 逐段找 SOFn
  if (buf[0] === 0xff && buf[1] === 0xd8) {
    let i = 2
    while (i + 9 < buf.length) {
      if (buf[i] !== 0xff) { i++; continue }
      const marker = buf[i + 1]
      if (marker === 0xd8 || marker === 0x01 || (marker >= 0xd0 && marker <= 0xd7)) { i += 2; continue }
      const len = buf.readUInt16BE(i + 2)
      // SOF0..SOF15，跳过 DHT(c4)/JPGA(c8)/DAC(cc)
      if (marker >= 0xc0 && marker <= 0xcf && marker !== 0xc4 && marker !== 0xc8 && marker !== 0xcc) {
        return { height: buf.readUInt16BE(i + 5), width: buf.readUInt16BE(i + 7) }
      }
      if (len < 2) return null
      i += 2 + len
    }
  }
  return null
}

export function isPanoramaRatio(size) {
  if (!size || !size.width || !size.height) return false
  const r = size.width / size.height
  return r >= PANO_RATIO_MIN && r <= PANO_RATIO_MAX
}

/** config 里的 panorama 路径 → exhibits/ 根相对路径（正斜杠）；远程 URL 返回 null */
export function panoramaToRootRel(exhibitDir, panoramaPath, exhibitsRoot) {
  const p = String(panoramaPath ?? '').trim()
  if (!p || isRemotePanoramaUrl(p)) return null
  let abs
  if (p.startsWith('/')) {
    if (!exhibitsRoot) return null
    abs = path.join(exhibitsRoot, p.replace(/^\/+/, ''))
  } else if (path.isAbsolute(p)) {
    abs = p
  } else {
    abs = path.join(exhibitDir, p)
  }
  if (!exhibitsRoot) return null
  const rel = path.relative(exhibitsRoot, abs)
  if (!rel || rel.startsWith('..') || path.isAbsolute(rel)) return null
  return rel.split(path.sep).join('/')
}

function readLocalImageMeta(exhibitsRoot, rootRel) {
  const full = path.join(exhibitsRoot, ...rootRel.split('/'))
  if (!PANO_EXT.has(path.extname(full).toLowerCase())) return null
  let fd = null
  const head = Buffer.alloc(65536)
  try {
    if (!fs.statSync(full).isFile()) return null
    fd = fs.openSync(full, 'r')
    const n = fs.readSync(fd, head, 0, head.length, 0)
    const size = imageSize(head.subarray(0, n))
    if (!isPanoramaRatio(size)) return null
    return { width: size.width, height: size.height }
  } catch {
    return null
  } finally {
    if (fd !== null) { try { fs.closeSync(fd) } catch {} }
  }
}

function addCandidate(map, entry) {
  if (map.size >= PANO_MAX_SCAN) return
  const prev = map.get(entry.path)
  if (!prev || (prev.source === 'shared' && entry.source === 'configured')) map.set(entry.path, entry)
}

function listExhibitDirs(exhibitsRoot) {
  const out = []
  let items
  try { items = fs.readdirSync(exhibitsRoot, { withFileTypes: true }) } catch { return out }
  for (const it of items) {
    if (!it.isDirectory() || it.name.startsWith('_') || it.name.startsWith('.')) continue
    try {
      if (fs.statSync(path.join(exhibitsRoot, it.name, 'config.json')).isFile()) out.push(it.name)
    } catch { /* 非展品目录 */ }
  }
  return out.sort()
}

/** 扫描可信来源，返回 [{ path, width, height, source }]（path 相对 exhibits/，正斜杠） */
export function listPanoramaCandidates(exhibitsRoot) {
  const root = path.resolve(exhibitsRoot)
  const map = new Map()

  for (const ex of listExhibitDirs(root)) {
    if (map.size >= PANO_MAX_SCAN) break
    const exhibitDir = path.join(root, ex)
    let cfg
    try { cfg = JSON.parse(fs.readFileSync(path.join(exhibitDir, 'config.json'), 'utf8')) } catch { continue }
    const pano = cfg?.assets?.panorama
    if (!pano || isRemotePanoramaUrl(pano)) continue
    const rootRel = panoramaToRootRel(exhibitDir, pano, root)
    if (!rootRel) continue
    const meta = readLocalImageMeta(root, rootRel)
    if (meta) addCandidate(map, { path: rootRel, ...meta, source: 'configured' })
  }

  const walkShared = (dir, rel) => {
    if (map.size >= PANO_MAX_SCAN) return
    let items
    try { items = fs.readdirSync(dir, { withFileTypes: true }) } catch { return }
    for (const it of items) {
      if (map.size >= PANO_MAX_SCAN) return
      if (it.name.startsWith('.')) continue
      const full = path.join(dir, it.name)
      const r = rel ? `${rel}/${it.name}` : it.name
      if (it.isDirectory()) {
        if (!PANO_SKIP_DIRS.has(it.name)) walkShared(full, r)
        continue
      }
      if (!it.isFile()) continue
      const meta = readLocalImageMeta(root, r)
      if (meta) addCandidate(map, { path: r, ...meta, source: 'shared' })
    }
  }

  for (const pub of PANO_PUBLIC_DIRS) {
    if (map.size >= PANO_MAX_SCAN) break
    const dir = path.join(root, ...pub.split('/'))
    try {
      if (fs.statSync(dir).isDirectory()) walkShared(dir, pub.replace(/\\/g, '/'))
    } catch { /* 目录不存在 */ }
  }

  const order = { configured: 0, shared: 1 }
  return [...map.values()].sort((a, b) =>
    (order[a.source] - order[b.source]) || a.path.localeCompare(b.path))
}
